nncf/common/graph/graph_matching.py

```diff
@@ -307,51 +45,57 @@ def find_whether_subgraph_has_inner_outgoing_edges(graph: nx.DiGraph, subgraph:
-        predecessors = list(graph.pred[node_key].keys())
-        for predecessors_key in predecessors:
-            if predecessors_key not in subgraph:
+        elif node_key == first_node:
+            successors = list(graph.succ[node_key].keys())
+            if any(successor not in subgraph for successor in successors):
+                return True
+        else:
+            successors = list(graph.succ[node_key].keys())
+            predecessors = list(graph.pred[node_key].keys())
+            if any(successors_key not in subgraph for successors_key in successors):
+                return True
+            if any(predecessor not in subgraph for predecessor in predecessors):
                 return True
     return False
 
 
-def find_subgraphs_matching_expression(graph: nx.DiGraph, expression: Expression) -> List[List[str]]:
+def find_subgraphs_matching_expression(graph: nx.DiGraph, pattern_graph: GraphPattern) -> List[List[str]]:
     """
     Find a list of subgraphs for the particular graph that match the pattern expression.
     :param graph: The model graph.
-    :param expression: A pattern expression containing a logic of layer fusing.
+    :param pattern_graph: A graph consists of patterns for layer fusing logic.
     :return: A list of subgraphs for the particular graph, matching the pattern expression.
     """
-    subgraphs = []
-    subgraphs_nodes = set()
-    nodes = nx.topological_sort(graph)
-    for node in nodes:
-        # If a node has already been added to any pattern skip this node
-        if node in subgraphs_nodes:
-            continue
 
-        all_matches = expression.all_matches([node], graph)
-        all_matches = sorted(all_matches, key=lambda x: len(x[0]), reverse=True)
+    def are_nodes_matching(node_1, node_2):
+        assert not isinstance(node_1['type'], list)
+        return node_1['type'] in node_2['type']
 
-        longest_valid_match = None
-        for match in all_matches:
-            # Find out the longest valid pattern
-            if not find_whether_subgraph_has_inner_outgoing_edges(graph, match[0]):
-                longest_valid_match = match
-                break
-        # If there is no pattern found, then skip this node
-        if longest_valid_match is None:
-            continue
-
-        for mn in longest_valid_match[0]:
-            subgraphs_nodes.add(mn)
+    subgraphs = []
+    visited_nodes = set()
+    patterns = []
+    for c in nx.weakly_connected_components(pattern_graph.graph):
+        patterns.append(pattern_graph.graph.subgraph(c))
+    # Get all patterns sorted by their lengths
+    # as we want match the longest patterns first
+    patterns = sorted(patterns, key=len, reverse=True)
+    for pattern in patterns:
+        matcher = ism.DiGraphMatcher(graph, pattern, node_match=are_nodes_matching)
+        for subgraph in matcher.subgraph_isomorphisms_iter():
+            subgraph = list(nx.topological_sort(graph.subgraph(subgraph)))
```

Comment:
@vshampor why is this a case? Firstly, I am not sure that first in topological order nodes must have smaller node counter than the following ones. Secondly, in lexicographical order node with counter '111' < '12', which is incorrect. So, I suggest remaining as it is.

Reply:
You may not leave it as it is because topological sorting does not give a unique result. If you need to implement special lexicographical portion of the sorting for your specific syntax of keys in the graph, use the corresponding interface of the `lexicographical_topological_sort` function.

Reply:
thank you!

Reply:
Done